share/p2p/shrex/server.go

```diff
@@ -0,0 +1,216 @@
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+const ndProtocolID = "/shrex/nd/0.0.1"
+
+var log = logging.Logger("shrex/nd")
+
+const (
+	writeTimeout = time.Second * 10
+	readTimeout  = time.Second * 10
+	serveTimeout = time.Second * 10
+)
+
+type Server struct {
+	store        *eds.Store
+	host         host.Host
+	writeTimeout time.Duration
+	readTimeout  time.Duration
+	serveTimeout time.Duration
+
+	// TODO: will be removed after CARBlockstore is ready to return DAH
+	testDAH *share.Root
+}
+
+func StartServer(host host.Host, store *eds.Store) *Server {
+	srv := &Server{
+		store:        store,
+		host:         host,
+		writeTimeout: writeTimeout,
+		readTimeout:  readTimeout,
+		serveTimeout: serveTimeout,
+	}
+	host.SetStreamHandler(ndProtocolID, srv.handleNamespacedData)
+	return srv
+}
+
+// Stop stops the server
+func (srv *Server) Stop() {
+	srv.host.RemoveStreamHandler(ndProtocolID)
+}
+
+func (srv *Server) handleNamespacedData(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(srv.readTimeout))
+	if err != nil {
+		log.Debugf("set read deadline: %s", err)
+	}
+
+	var req share_p2p_v1.GetSharesByNamespaceRequest
+	_, err = serde.Read(stream, &req)
+	if err != nil {
+		log.Errorw("read msg", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		log.Debugf("close read side of the stream: %s", err)
+	}
+
+	err = validateRequest(req)
+	if err != nil {
+		log.Errorw("invalid request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), srv.serveTimeout)
+	defer cancel()
+
+	bs, dah, err := srv.store.CARBlockstore(ctx, req.RootHash)
+	if err != nil {
+		log.Errorw("get header for datahash", "err", err)
+		srv.respondInternal(stream)
+		return
+	}
+
+	// replace dah with testDAH within test scope
+	if srv.testDAH != nil {
+		dah = srv.testDAH
+	}
+
+	getter := eds.NewBlockGetter(bs)
+	blob, err := getBlobByNamespace(ctx, getter, dah, req.NamespaceId)
+	if err != nil {
+		log.Errorw("get shares", "err", err)
+		srv.respondInternal(stream)
+		return
+	}
+
+	resp := blobToResponse(blob)
+	srv.respond(stream, resp)
+}
+
+func getBlobByNamespace(
+	ctx context.Context,
+	getter blockservice.BlockGetter,
+	root *share.Root,
+	nID namespace.ID,
+) (*share.Blob, error) {
+	// select rows that contain shares with desired namespace.ID
+	selectedRows := make([]cid.Cid, 0)
+	for _, row := range root.RowsRoots {
+		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
+			selectedRows = append(selectedRows, ipld.MustCidFromNamespacedSha256(row))
+		}
```

Comment:
I believe this already exists as a helper function, and if it doesn't yet it should